enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -6131,39 +6411,72 @@ class AdjointGenerator
-          dx = Builder2.CreateFMul(dx, diffe(orig, Builder2));
-          setDiffe(orig, Constant::getNullValue(orig->getType()), Builder2);
-          addToDiffe(orig->getArgOperand(1), dx, Builder2, x->getType());
-          return;
+          switch (Mode) {
+          case DerivativeMode::ForwardMode: {
+            IRBuilder<> Builder2(&call);
+            getForwardBuilder(Builder2);
+            Value *x = gutils->getNewFromOriginal(orig->getArgOperand(1));
+            Value *n = gutils->getNewFromOriginal(orig->getArgOperand(0));
+
+            Value *d0 = Builder2.CreateCall(
+                called,
+                std::vector<Value *>(
+                    {Builder2.CreateSub(n, ConstantInt::get(n->getType(), 1)),
+                     x}));
+
+            Value *d2 = Builder2.CreateCall(
+                called,
+                std::vector<Value *>(
+                    {Builder2.CreateAdd(n, ConstantInt::get(n->getType(), 1)),
+                     x}));
+
+            Value *dx = Builder2.CreateFSub(d0, d2);
+            dx = Builder2.CreateFMul(dx, ConstantFP::get(x->getType(), 0.5));
+            dx = Builder2.CreateFMul(dx,
+                                     diffe(orig->getArgOperand(1), Builder2));
+            setDiffe(orig, dx, Builder2);
+            return;
+          }
+          case DerivativeMode::ReverseModeGradient:
+          case DerivativeMode::ReverseModeCombined: {
+            IRBuilder<> Builder2(call.getParent());
+            getReverseBuilder(Builder2);
+            Value *x = lookup(
+                gutils->getNewFromOriginal(orig->getArgOperand(1)), Builder2);
+            Value *n = lookup(
+                gutils->getNewFromOriginal(orig->getArgOperand(0)), Builder2);
+
+            Value *d0 = Builder2.CreateCall(
+                called,
+                std::vector<Value *>(
+                    {Builder2.CreateSub(n, ConstantInt::get(n->getType(), 1)),
+                     x}));
+
+            Value *d2 = Builder2.CreateCall(
+                called,
+                std::vector<Value *>(
+                    {Builder2.CreateAdd(n, ConstantInt::get(n->getType(), 1)),
+                     x}));
+
+            Value *dx = Builder2.CreateFSub(d0, d2);
+            dx = Builder2.CreateFMul(dx, ConstantFP::get(x->getType(), 0.5));
+            dx = Builder2.CreateFMul(dx, diffe(orig, Builder2));
+            setDiffe(orig, Constant::getNullValue(orig->getType()), Builder2);
+            addToDiffe(orig->getArgOperand(1), dx, Builder2, x->getType());
+            return;
+          }
+          case DerivativeMode::ReverseModePrimal: {
+            return;
+          }
+          }
         }
 
         if (funcName == "julia.write_barrier") {
-          if (Mode == DerivativeMode::ReverseModeGradient) {
+          if (Mode == DerivativeMode::ReverseModeGradient ||
+              Mode == DerivativeMode::ForwardMode) {
             eraseIfUnused(*orig, /*erase*/ true, /*check*/ false);
```

Comment:
The write barrier needs to also be run on shadow pointers on the forward (as below)